notebooks/03-query/15-pivot-unpivot.ipynb

```diff
@@ -0,0 +1,372 @@
+    "'''\n",
+    "display_client.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a8642f39-ad04-45b3-9120-b622a0f7c8a7",
+   "metadata": {},
+   "source": [
+    "## PIVOT example\n",
+    "The prior result is a bit long and makes it hard to figure out which channels have the most updates while still seeing the distinction between additions and deletions to the wikipedia pages.\n",
+    "One way to reduce the result while still keeping all the detail is to pivot the result such that rows become columns for each value of a particular column, the pivot column.\n",
+    "\n",
+    "PIVOT is a SQL operator that is used after the FROM clause.\n",
+    "\n",
+    "```\n",
+    "    FROM \"example-pivot-unpivot\" \n",
+    "    \n",
+    "    PIVOT ( SUM(added) as added, \n",
+    "            SUM(deleted) as deleted \n",
+    "            \n",
+    "            FOR \"isRobot\" IN ('true' as robot, 'false' as human)\n",
+    "          )\n",
+    "```\n",
+    "- It lists a set of aggregations to calculate: `SUM(added) as added, SUM(deleted) as deleted`\n",
+    "- The FOR clause indicated the pivot column: `\"isRobot\"`\n",
+    "- The IN clause lists the pivot values of the column and what to call them: `IN ('true' as robot, 'false' as human)`\n",
+    "\n",
+    "The resulting columns from this operation are a combination in the form \\<pivot_value_name\\>__\\<aggregation_name\\>. So in this case the result pivot columns will be:\n",
+    "```\n",
+    "\"robot_added\"\n",
+    "\"human_added\"\n",
+    "\"robot_deleted\"\n",
+    "\"human_deleted\"\n",
+    "```\n",
+    "which can now form part of the SELECT clause in the query.\n",
+    "\n",
+    "\n",
+    "The following query uses the results of the pivot to then aggregate totals for the pivot columns by channel. The aggregations being calculated could be NULL in the case where \"isRobot\"='true' (or 'false') does not exist for a given channel. The query calculates a ratio of human to total changes for each channel and the result is sorted on this ratio such that the channels with the highest proportion of robot updates are listed first. So in order to calculate this metric even in presence of NULLs the query uses COALESCE on all the pivoted metrics.\n",
+    "\n",
+    "The result is much cleaner than the prior query and it is easy to determine which channels have the most robotic activity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "65c489a8-a7f2-4cc0-b30c-f30d8ed32be2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT \"channel\", \n",
+    "        SUM(\"robot_added\")       AS \"added_robot\", \n",
+    "        SUM(\"robot_deleted\")     AS \"deleted_robot\", \n",
+    "        SUM(\"human_added\")       AS \"added_human\",\n",
+    "        SUM(\"human_deleted\")     AS \"deleted_human\", \n",
+    "        \n",
+    "        SAFE_DIVIDE( SUM(\"human_deleted\" + \"human_added\") * 1.0 , \n",
```

Comment:
SAFE_DIVIDE Is cool here! I don't think we have a notebook on this function, so I'd suggest you call out this particularly, with a link to the docs on this function, and why it forms part of this query.

Reply:
done.